server/src/main/java/com/github/phantazmnetwork/server/Messaging.java

```diff
@@ -0,0 +1,58 @@
+package com.github.phantazmnetwork.server;
+
+import com.github.phantazmnetwork.commons.Namespaces;
+import com.github.phantazmnetwork.core.player.BasicPlayerViewProvider;
+import com.github.phantazmnetwork.messaging.MessageChannels;
+import com.github.phantazmnetwork.messaging.packet.PacketSerializer;
+import com.github.phantazmnetwork.messaging.packet.PacketSerializers;
+import com.github.phantazmnetwork.messaging.proxy.ForwardProtocolVersionPacket;
+import com.github.phantazmnetwork.server.packet.BinaryDataReader;
+import com.github.phantazmnetwork.server.packet.BinaryDataWriter;
+import net.minestom.server.event.Event;
+import net.minestom.server.event.EventNode;
+import net.minestom.server.event.player.PlayerDisconnectEvent;
+import net.minestom.server.event.player.PlayerPluginMessageEvent;
+import net.minestom.server.utils.binary.BinaryReader;
+import net.minestom.server.utils.binary.BinaryWriter;
+import org.jetbrains.annotations.NotNull;
+
+public final class Messaging {
+
+    private Messaging() {
+        throw new UnsupportedOperationException();
+    }
+
+    static void initialize(@NotNull EventNode<Event> global, @NotNull BasicPlayerViewProvider basicPlayerViewProvider) {
+        PacketSerializer proxyPacketSerializer =
+                PacketSerializers.createProxySerializer(() -> new BinaryDataWriter(new BinaryWriter()),
+                                                        bytes -> new BinaryDataReader(new BinaryReader(bytes))
+                );
+
+        global.addListener(PlayerPluginMessageEvent.class, event -> {
+            String identifier = event.getIdentifier();
+            if (!identifier.contains(":")) {
+                return;
+            }
+
+            String[] split = identifier.split(":");
+            if (split.length != 2 || !split[0].equals(Namespaces.PHANTAZM)) {
+                return;
+            }
+
+            if (split[1].equals(MessageChannels.PROXY)) {
+                proxyPacketSerializer.deserializePacket(event.getMessage()).ifPresent(packet -> {
```

Comment:
I do not believe it is safe
we can intercept plugin messages that are sent from the client to the server since that goes through the proxy first to see if it's sent on the proxy channel
I don't believe clients can send general packets directly, they have to do it through the proxy

Reply:
it would likely be trivial to sign these packets with a shared secret

Reply:
a player running their own instance of velocity (and with some custom code probably) could in theory send version change packets from that to bypass version restrictions

Reply:
the player would have to connect in the first place though

Reply:
is ServerPostConnectEvent fired even if the player is blocked by BungeeGuard?

Reply:
no